providers/common/der/der_rsa_key.c

```diff
@@ -375,8 +372,29 @@ int ossl_DER_w_algorithmIdentifier_RSA(WPACKET *pkt, int tag, RSA *rsa)
 
     return ossl_DER_w_begin_sequence(pkt, tag)
         && (rsa_nid != NID_rsassaPss
-            || ossl_rsa_pss_params_30_is_unrestricted(pss_params)
-            || ossl_DER_w_RSASSA_PSS_params(pkt, -1, pss_params))
+            || ossl_rsa_pss_params_30_is_unrestricted(pss)
+            || ossl_DER_w_RSASSA_PSS_params(pkt, -1, pss))
         && ossl_DER_w_precompiled(pkt, -1, rsa_oid, rsa_oid_sz)
         && ossl_DER_w_end_sequence(pkt, tag);
 }
+
+int ossl_DER_w_algorithmIdentifier_RSA(WPACKET *pkt, int tag, RSA *rsa)
+{
+    int rsa_nid = NID_undef;
+    RSA_PSS_PARAMS_30 *pss_params = ossl_rsa_get0_pss_params_30(rsa);
+
+    switch (RSA_test_flags(rsa, RSA_FLAG_TYPE_MASK)) {
+    case RSA_FLAG_TYPE_RSA:
+        rsa_nid = NID_rsaEncryption;
+        break;
+    case RSA_FLAG_TYPE_RSASSAPSS:
+        rsa_nid = NID_rsassaPss;
+        break;
+    }
+
+    if (rsa_nid == NID_undef)
+        return 0;
+
+    return ossl_DER_w_algorithmIdentifier_RSA_params(pkt, tag, rsa_nid,
+                                                     pss_params);
+}
```

Comment:
This could also pass the RSA type directly, like this:

``` C
int ossl_DER_w_algorithmIdentifier_RSA(WPACKET *pkt, int tag, RSA *rsa)
{
    RSA_PSS_PARAMS_30 *pss_params = ossl_rsa_get0_pss_params_30(rsa);
    int rsa_type = RSA_test_flags(rsa, RSA_FLAG_TYPE_MASK);

    return ossl_DER_w_algorithmIdentifier_RSA_params(pkt, tag, rsa_type,
                                                     pss_params);
}
```

Then you need to make even fewer changes to `ossl_DER_w_algorithmIdentifier_RSA_params()`